sys/net/gnrc/link_layer/gnrc_mac/internal.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (C) 2015 Daniel Krebs
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net_gnrc_mac
+ * @file
+ * @brief       Internal functions of GNRC_WMAC
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @author      Shuguo Zhuo  <shuguo.zhuo@inria.fr>
+ * @}
+ */
+
+#include <stdbool.h>
+#include <net/gnrc.h>
+#include <net/gnrc/mac/internal.h>
+#include <net/gnrc/mac/types.h>
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+
+#if ((GNRC_MAC_TX_QUEUE_SIZE != 0)||(GNRC_MAC_RX_QUEUE_SIZE != 0))
+gnrc_priority_pktqueue_node_t* _alloc_pktqueue_node(gnrc_priority_pktqueue_node_t* nodes[], uint32_t size)
+{
+    /* search for free packet_queue_node */
+    for (size_t i = 0; i < size; i++) {
+        if((nodes[i]->pkt == NULL) &&
+           (nodes[i]->next == NULL)) {
+            return nodes[i];
+        }
+    }
+
+    return NULL;
+}
+#endif
+
+#if GNRC_MAC_TX_QUEUE_SIZE != 0
+#if GNRC_MAC_NEIGHBOUR_COUNT != 0
+int _find_neighbour(gnrc_mac_tx_t* tx, uint8_t* dst_addr, int addr_len)
+{
+    gnrc_mac_tx_neighbour_t* neighbours;
+    neighbours = tx->neighbours;
+
+    for(int i = 0; i <= GNRC_MAC_NEIGHBOUR_COUNT; i++) {
+        if(neighbours[i].l2_addr_len == addr_len) {
+            if(memcmp(&(neighbours[i].l2_addr), dst_addr, addr_len) == 0) {
+                return i;
+            }
+        }
+    }
+    return -1;
+}
+
+/* Free first empty queue that is not active */
+int _free_neighbour(gnrc_mac_tx_t* tx)
+{
+    gnrc_mac_tx_neighbour_t* neighbours;
+    neighbours = tx->neighbours;
+
+    /* Don't attempt to free broadcast neighbour, so start at index 1 */
+    for(int i = 1; i <= GNRC_MAC_NEIGHBOUR_COUNT; i++) {
+        if((gnrc_priority_pktqueue_length(&(neighbours[i].queue)) == 0) &&
+           (&neighbours[i] != tx.current_neighbour)) {
+            /* Mark as free */
+            neighbours[i].l2_addr_len = 0;
+            return i;
+        }
+    }
+    return -1;
+}
+
+int _alloc_neighbour(gnrc_mac_tx_t* tx)
+{
+    gnrc_mac_tx_neighbour_t* neighbours;
+    neighbours = tx->neighbours;
+
+    for(int i = 0; i <= GNRC_MAC_NEIGHBOUR_COUNT; i++) {
+        if(neighbours[i].l2_addr_len == 0) {
+            gnrc_priority_pktqueue_init(&(neighbours[i].queue));
+            return i;
+        }
+    }
+    return -1;
+}
+
+void _init_neighbour(gnrc_mac_tx_neighbour_t* neighbour, uint8_t* addr, int len)
+{
+    assert(neighbour != NULL);
+    assert(addr  != NULL);
+    assert(len > 0);
+
+    neighbour->l2_addr_len = len;
+    neighbour->phase = GNRC_MAC_PHASE_UNINITIALIZED;
+    memcpy(&(neighbour->l2_addr), addr, len);
+}
+#endif
```

Comment:
Please add here and for all following `... #else ... #endif` a comment which `#if` it belongs to.